johnnydep/pipper.py

```diff
@@ -75,6 +75,19 @@ def _get_wheel_args(index_url, env, extra_index_url):
     return args
 
 
+def _download_dist(url, scratch_file, index_url, extra_index_url):
+    if index_url and ":" in index_url and "@" in index_url and urlparse(url).netloc in index_url:
+        # handling private PyPI credentials in index_url
+        auth = tuple(urlparse(index_url).netloc.split("@")[0].split(":"))
+    elif extra_index_url and ":" in extra_index_url and "@" in extra_index_url and urlparse(url).netloc in extra_index_url:
+        # handling private PyPI credentials in extra_index_url
+        auth = tuple(urlparse(extra_index_url).netloc.split("@")[0].split(":"))
+    else:
+        auth = None
+    target, _headers = urlretrieve(url, scratch_file, auth=auth)
+    return target, _headers
```

Comment:
There is a potential security vulnerability here.  If both `--index-url` and `--extra-index-url` were specified (different indexes), then the user/pass from any url will be sent to *all* of the urls.  The auth should only really be used if the `url` in argument is matching the index that needs auth.

Reply:
the last `and` clauses (`urlparse(url).netloc in index_url` and `urlparse(url).netloc in extra_index_url`) prevent  the wrong index credentials being used as auth for `url`. we can also assume non-identical index urls, and if they were identical it wouldnt be an issue as the auth of index_url would be used to request from url (which has the same netloc). this is also covered by the test [`"extra_index_url with auth"`](https://github.com/wimglenn/johnnydep/blob/df95ab6cba132bbcfbb66d523d7539c553c6cb6e/tests/test_pipper.py#L82-L85)

Reply:
Hmm, I see.  It's not as bad as I thought, but `urlparse(url).netloc in index_url` is not really the correct way to do that, since this substring will check the whole url including path.  If one url was `https://user:pass@alice.org/` you don't want to send credentials to "https://eve.org/alice.org/", right?

Reply:
Oh, one other thing, the check for presence of credentials should be in netloc part only - and you should get the pieces with attributes not string ops:

```
parsed = urlparse("https://u:p@www.example.org/")
user = parsed.username
pwd = parsed.password

```